source/extensions/common/wasm/wasm_vm.h

```diff
@@ -0,0 +1,199 @@
+#pragma once
+
+#include <memory>
+
+#include "envoy/common/exception.h"
+
+#include "common/common/logger.h"
+
+#include "absl/types/optional.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Common {
+namespace Wasm {
+
+class Context;
+
+// Represents a WASM-native word-sized datum. On 32-bit VMs, the high bits are always zero.
+// The WASM/VM API treats all bits as significant.
+struct Word {
+  Word(uint64_t w) : u64(w) {}              // Implicit conversion into Word.
+  operator uint64_t() const { return u64; } // Implicit conversion into uint64_t.
+  // Note: no implicit conversion to uint32_t as it is lossy.
+  uint32_t u32() const { return static_cast<uint32_t>(u64); }
+  uint64_t u64;
+};
+
+template <typename T> struct ConvertWordTypeToUint32 { using type = T; };
+template <> struct ConvertWordTypeToUint32<Word> { using type = uint32_t; };
+
+template <typename F> struct ConvertFunctionTypeWordToUint32 {};
+template <typename R, typename... Args> struct ConvertFunctionTypeWordToUint32<R (*)(Args...)> {
+  using type = typename ConvertWordTypeToUint32<R>::type (*)(
+      typename ConvertWordTypeToUint32<Args>::type...);
+};
+
+template <typename T> struct Global {
+  virtual ~Global() {}
+  virtual T get() PURE;
+  virtual void set(const T& t) PURE;
+};
+
+// Calls into the WASM VM.
+// 1st arg is always a pointer to Context (Context*).
+using WasmCall0Void = std::function<void(Context*)>;
```

Comment:
Can't this be templatized with (Args...)?

Reply:
The explicit instantiations are required in order to get the virtual function prototypes.  The original implementation used templates but because template virtual functions are not legal in C++, it needed to use regular function templates and have the different VM implementation versions exposed (no virtuals).   Despite the fact that this simplified the code, @PiotrSikora did not like it (I believe his argument was that it would be hard to upstream) and so we changed it to what you see here.  Personally I would go back to the templates, and it would be lovely irony you wanted that for upstreaming (hint hint).

Reply:
The reason for that change (https://github.com/envoyproxy/envoy-wasm/commit/0e61789a57c9f0c0e8f5d430207fa69fc0e34da1) wasn't "hard to upstream", but the fact that the implementation details were leaking into the common abstraction layer.

Reply:
Yes and that prevented use from using a registration system.  As Dirk Gently would say "it is all interrelated".

Reply:
I didn't mean to use template for registration, but just for those generated types with template magic.. I got something working:
https://gist.github.com/lizan/d9fd1242c96218a0c590c7013959be9d